cmd/metal-api/internal/grpc/boot-service-wait_integration_test.go

```diff
@@ -5,343 +5,341 @@ package grpc
-	}
+	var (
+		allocations = make(chan string)
+		machines    []string
+	)
 
-	ds, mock := datastore.InitMockDB(t.T)
-	for i := range t.numberMachineInstances {
-		machineID := strconv.Itoa(i)
-		mock.On(r.DB("mockdb").Table("machine").Get(machineID)).Return(metal.Machine{Base: metal.Base{ID: machineID}}, nil)
-		mock.On(insertMock(true, machineID)).Return(returnMock(true, machineID), nil)
-		mock.On(insertMock(false, machineID)).Return(returnMock(false, machineID), nil)
-	}
+	te.startApiInstances(ctx)
+	te.startMachineInstances(ctx, allocations)
 
-	t.startApiInstances(ds)
-	t.startMachineInstances()
-	t.notReadyMachines.Wait()
+	for i := range te.numberAllocations {
+		client := te.cc[i%len(te.cc)]
+		te.t.Logf("sending nsq allocation event for machine %s", client.machineID)
 
-	require.Len(t, wait, t.numberMachineInstances)
-	for _, wait := range wait {
-		require.True(t, wait)
-	}
+		err := te.publisher.Publish(metal.TopicAllocation.Name, &metal.AllocationEvent{MachineID: client.machineID})
+		require.NoError(te.t, err)
 
-	switch t.testCase {
-	case happyPath:
-	case serverFailure:
-		t.notReadyMachines.Add(t.numberMachineInstances)
-		t.stopApiInstances()
-		t.startApiInstances(ds)
-		t.notReadyMachines.Wait()
-	case clientFailure:
-		t.notReadyMachines.Add(t.numberMachineInstances)
-		t.stopMachineInstances()
-		t.startMachineInstances()
-		t.notReadyMachines.Wait()
+		machines = append(machines, client.machineID)
 	}
 
-	require.Len(t, wait, t.numberMachineInstances)
-	for _, wait := range wait {
-		require.True(t, wait)
-	}
+	waitCtx, waitCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer waitCancel()
 
-	t.allocateMachines()
+	for {
+		select {
+		case <-waitCtx.Done():
+			te.t.Errorf("not all machines left wait for allocation loop in time")
+		case machineID := <-allocations:
+			require.Contains(te.t, machines, machineID, "unexpected machine left wait loop")
 
-	t.unallocatedMachines.Wait()
+			te.t.Logf("machine %q left wait for allocation loop", machineID)
 
-	require.Len(t, t.allocations, t.numberAllocations)
-	for _, allocated := range t.allocations {
-		require.True(t, allocated)
-	}
+			machines = slices.DeleteFunc(machines, func(id string) bool {
+				return id == machineID
+			})
+		}
 
-	require.Len(t, wait, t.numberMachineInstances)
-	for key, wait := range wait {
-		require.Equal(t, !containsKey(t.allocations, key), wait)
+		if len(machines) == 0 {
+			te.t.Logf("all expected machines left wait for allocation loop")
+			break
+		}
 	}
-}
 
-func containsKey(m map[string]bool, key string) bool {
-	for k := range m {
-		if k == key {
-			return true
+	ms, err := te.ds.ListMachines()
+	require.NoError(te.t, err)
+
+	waiting := 0
+	for _, m := range ms {
+		if m.Waiting {
+			waiting++
 		}
 	}
-	return false
-}
 
-func (t *test) shutdown() {
-	t.stopMachineInstances()
-	t.stopApiInstances()
+	assert.Equal(te.t, te.numberMachineInstances-te.numberAllocations, waiting)
+	assert.Len(te.t, ms, te.numberMachineInstances)
```

Comment:
If we are really ambitious, we could use the remaining clients which are waiting for testing an "outage scenario", i.e. shutting down the grpc-servers, check that the clients do not give up and reconnect after bringing the servers back up (and vice versa).

Reply:
lets try this

Reply:
Done.